implement Observer pattern + Updater & Requirements

weedySeaDragon opened the pull request:
## PT Story:  User Observer pattern to update membership status
https://www.pivotaltracker.com/story/show/154431789

## Changes proposed in this pull request:
1. Use the Observer pattern to notify the MembershipUpdater if changes were made that affect the membership status
2. implemented an Updater that checks with Requirements pattern so that we can  put all requirements in one place (ex: all requirements for granting membership)
3. See the write-up in the wiki: 
 **[Refactoring membership status management to use the Observer Pattern (PR 445 write up)](https://github.com/AgileVentures/shf-project/wiki/Refactoring-membership-status-management-to-use-the-Observer-Pattern---PR-445-write-up)**

I took the (spike) PR #445  and cleaned it up a bit and applied it to the current development branch.

I did add a .feature file that is there so we can specifically test payment + date use cases.  
I like having those tests in one place, but I could also split them up and put them into existing .feature files that address similar areas.

Ready for review:
@patmbolger

tomurb committed:
set default from, reply-to and display name

tomurb committed:
skipped earlier

tomurb committed:
Merge branch 'develop' into sender-name-of-emails#154355653

tomurb committed:
deleted accidental addition

weedySeaDragon committed:
Merge branch 'sender-name-of-emails#154355653' of https://github.com/tomurb/shf-project into develop

weedySeaDragon committed:
Merge branch 'develop' of https://github.com/AgileVentures/shf-project into develop

weedySeaDragon committed:
add AbstractUpdater class

weedySeaDragon committed:
add AbstractUpdater class

weedySeaDragon committed:
User: add has_approved_shf_application? (req'd for MembershipRequirements)

weedySeaDragon committed:
User: add has_approved_shf_application? (req'd for MembershipRequirements)

weedySeaDragon committed:
Payment factory: add traits, factory types for easier testing

weedySeaDragon committed:
Payment factory: add traits, factory types for easier testing

weedySeaDragon committed:
User factory: comments: FIXME: don't add a list of applications, just add 1

weedySeaDragon committed:
User factory: comments: FIXME: don't add a list of applications, just add 1

weedySeaDragon committed:
Add AbstractRequirements class

weedySeaDragon committed:
Add AbstractRequirements class

weedySeaDragon committed:
add RequirementsForMembership, RequirementsForRevokingMembership classes

weedySeaDragon committed:
add RequirementsForMembership, RequirementsForRevokingMembership classes

weedySeaDragon committed:
add MembershipStatusUpdater class

weedySeaDragon committed:
add MembershipStatusUpdater class

weedySeaDragon committed:
update requirements.puml with renamed .has_expected_arguments? method name

weedySeaDragon committed:
update requirements.puml with renamed .has_expected_arguments? method name

weedySeaDragon committed:
simple class diagram of MembershipStatusManager +User +Payments

weedySeaDragon committed:
simple class diagram of MembershipStatusManager +User +Payments

weedySeaDragon committed:
remove membership status responsibility from User

weedySeaDragon committed:
remove membership status responsibility from User

weedySeaDragon committed:
+Observable to Payment; add constants

weedySeaDragon committed:
+Observable to Payment; add constants

weedySeaDragon committed:
remove membership responsibility from Payment, PaymentController

weedySeaDragon committed:
remove membership responsibility from Payment, PaymentController

weedySeaDragon committed:
add Observer to ShfApplication

weedySeaDragon committed:
add Observer to ShfApplication

weedySeaDragon committed:
after sign in, MembershipUpdater checks status of user

weedySeaDragon committed:
after sign in, MembershipUpdater checks status of user

weedySeaDragon committed:
Seed Helpers: update to use MembershipUpdater - IN PROGRESS

weedySeaDragon committed:
Seed Helpers: update to use MembershipUpdater - IN PROGRESS

weedySeaDragon committed:
update Membership status in seeding (when application == accepted)

weedySeaDragon committed:
update Membership status in seeding (when application == accepted)

weedySeaDragon committed:
add helper method branding_fee? to Payment

weedySeaDragon committed:
add helper method branding_fee? to Payment

weedySeaDragon committed:
payment.successfully_completed so observers are notified; refactor a bit

weedySeaDragon committed:
payment.successfully_completed so observers are notified; refactor a bit

weedySeaDragon committed:
move test to create_order: was membership updated correctly?

weedySeaDragon committed:
move test to create_order: was membership updated correctly?

weedySeaDragon committed:
tiny comment edit

weedySeaDragon committed:
tiny comment edit

weedySeaDragon committed:
call payment.successfully_completed only when sucess action is called

weedySeaDragon committed:
call payment.successfully_completed only when sucess action is called

weedySeaDragon committed:
updated sequence diagram for observer/updater/requirements

weedySeaDragon committed:
updated sequence diagram for observer/updater/requirements

weedySeaDragon committed:
add date ParameterType to make steps easier

weedySeaDragon committed:
add date ParameterType to make steps easier

weedySeaDragon committed:
remove test from PaymentController#create

weedySeaDragon committed:
remove test from PaymentController#create

weedySeaDragon committed:
steps to improve readability and check membership status, expiration

weedySeaDragon committed:
steps to improve readability and check membership status, expiration

weedySeaDragon committed:
remove unneeded methods; correct CREATED key

weedySeaDragon committed:
remove unneeded methods; correct CREATED key

weedySeaDragon committed:
add specs for Payment#successfully_completed

weedySeaDragon committed:
add specs for Payment#successfully_completed

weedySeaDragon committed:
improve/refine Updaters

weedySeaDragon committed:
improve/refine Updaters

weedySeaDragon committed:
standardize log file name

weedySeaDragon committed:
standardize log file name

weedySeaDragon committed:
fix error

weedySeaDragon committed:
fix error

weedySeaDragon committed:
add .feature to test various payment + date scenarios

weedySeaDragon committed:
add .feature to test various payment + date scenarios

patmbolger commented on app/models/abstract_updater.rb:
```diff
@@ -0,0 +1,152 @@
+
+
+  # It is the responsibility of each subclass to parse the arguments as needed.
+  # They are an array here to allow for variation in the number and type(s) of
```
' ... an array ... ' >> ' ... a hash ... '

patmbolger commented on app/models/abstract_requirements.rb:
```diff
@@ -0,0 +1,52 @@
+#
+#       Each subclass MUST define the following methods:
+#
+#        'self.has_expected_arguments??(args)' verifies that the all the arguments (a Hash)
```
'??' >> '?'

patmbolger commented on app/models/membership_status_updater.rb:
```diff
@@ -0,0 +1,160 @@
+#      1. include Observable
+#      2. after_initialization :add_observers
+#      3. def add_observers (and add the observers in that method)
+#      4. where appropriate, let the observerse know a change has been made so they can do their thing
```
'observers' mispelled.

patmbolger commented on features/membership_status/member-status-change.feature:
```diff
@@ -0,0 +1,101 @@
+  # --- MEMBERSHIP PAYMENTS -------------
+
+  @time_adjust
+  Scenario: Membership payment made before membership expires
```
Need a "Given the date is ...." initial step to ensure that the test does not fail after 2018-12-31.

weedySeaDragon replied:
After merging with the current development, I'm having trouble getting this to pass.

Does it pass for you?

(and after #562 is merged, this will change)

patmbolger commented on spec/controllers/payments_controller_spec.rb:
```diff
@@ -53,6 +53,11 @@
+    let(:valid_order) do
+      payment_data = { id: 1, type: Payment::PAYMENT_TYPE_MEMBER, currency: 'SEK' }
+      HipsService.create_order(1, 1, payment_data, nil_urls)
+    end
```
This test object is not needed in this spec.

patmbolger commented on spec/controllers/payments_controller_spec.rb:
```diff
@@ -70,18 +75,4 @@
-      expect(payment.reload.user.member).to be true
-      expect(payment.user.membership_number).not_to be_nil
-    end
-  end
```
I think this test in useful to retain as it confirms that the controller action - upon payment success - including triggering the membership status update mechanism.

The test description should be changed to something like: 'triggers membership update mechanism'.

What do you think?

weedySeaDragon replied:
Yes --  we could have the test be something like:
```
it 'calls payment.successfully_completed ' do
  < expect that payment receives :successfully_completed message>
end
```
Doing more than that seems to me that we're then testing Observable.

patmbolger reviewed (changes requested):
Looks good.  Please see comments.

weedySeaDragon committed:
fix typos

weedySeaDragon committed:
add spec for PaymentController: on success, payment.payment_successfully_completed is called

weedySeaDragon committed:
show info about the user if 'should be a member' expectation fails

weedySeaDragon committed:
set the date at the start of the scenario

weedySeaDragon committed:
Merge remote-tracking branch 'origin/sprint54-Observer-Updaters-Requirements' into sprint54-Observer-Updaters-Requirements

# Conflicts:
#	app/mailers/application_mailer.rb
#	app/models/abstract_requirements.rb
#	app/models/abstract_updater.rb
#	app/models/membership_status_updater.rb
#	features/emails/email-applicant-membership-granted.feature
#	features/emails/email-applicant-when-app-approved.feature
#	features/emails/new-application-email-to-admins.feature
#	features/membership_status/member-status-change.feature
#	features/step_definitions/membership_status_steps.rb
#	spec/controllers/payments_controller_spec.rb
#	spec/mailers/application_mailer_spec.rb
#	spec/mailers/shared_email_tests.rb

weedySeaDragon commented:
oh bother.  I blorfed this up by merging/rebasing nonsense.  <sigh>

Once things are approved I'll create a new clean branch so I don't mess up the main /develop branch with all of this nonsense.

@patmbolger  I fixed the typos and added a PaymentsController test.

I also cannot get the features/membership_status/member-status-change.feature   scenarios to pass.  I know that the results will be different once your #562 is merged.  Even so, they're not working.  
You can wait to look at those features until after I do a new, cleaner, PR if you'd like.

weedySeaDragon committed:
Merge branch 'AV-develop' into sprint54-Observer-Updaters-Requirements

# Conflicts:
#	features/step_definitions/payment_steps.rb
#	spec/models/user_spec.rb

weedySeaDragon committed:
fix mailer model to get receipient email

weedySeaDragon committed:
comment out specs that are now part of MembershipUpdater

weedySeaDragon committed:
add mailing changes from PR 556 that got clobbered during merges

weedySeaDragon committed:
update steps to "And I complete the membership payment"

weedySeaDragon commented:
@patmbolger The first scenario in `features/membership_status/member-status-change.feature` is still not passing.  I see in your (updated) step you say that the payments must be created so that the 'created_at' date is set _correctly_ because that is what's used for ordering the payments and getting the last payment & expiration date.  (sorry for the run-on sentence)

How should I do that so that the scenarios pass?  Do I need to have a `Given the date is...X` statement up in the `Background` section? And set that date to anything before the dates set in the scenarios?

weedySeaDragon committed:
Set the date in the Background section so payment created_at dates are sooner

weedySeaDragon commented:
@patmbolger  changes were made and I updated `features/membership_status/member-status-change.feature` scenarios.  I added a `Given the date is...` into the Background section so that the created_at dates were early enough (and so would be sorted correctly).

weedySeaDragon committed:
Merge branch 'AV-develop' into sprint54-Observer-Updaters-Requirements
